src/spurt/workflows/emcf/_solver.py

```diff
@@ -0,0 +1,266 @@
+Phase Unwrapping of Multitemporal Differential SAR Interferograms," in IEEE
+Transactions on Geoscience and Remote Sensing, vol. 44, no. 9, pp. 2374-2383,
+Sept. 2006, doi: 10.1109/TGRS.2006.873207.
+"""
+
+from __future__ import annotations
+
+import numpy as np
+
+from spurt.links import LinkModelInterface
+from spurt.mcf import MCFSolverInterface, utils
+
+from ._settings import Settings
+
+
+class EMCFSolver:
+    """Implementation of the EMCF algorithm."""
+
+    def __init__(
+        self,
+        solver_space: MCFSolverInterface,
+        solver_time: MCFSolverInterface,
+        settings: Settings | None = None,
+        link_model: LinkModelInterface | None = None,
+    ):
+        """Spatio-temporal unwrapping.
+
+        Parameters
+        ----------
+        solver_space: MCFSolverInterface
+            MCF Solver interface for spatial graph connecting the stable points.
+            Typically a Delaunaytriangulation.
+        solver_time: MCFSolverInterface
+            MCF Solver interface for temporal graph usually representing
+            interferograms in time-Bperp space. Typically a Delaunay triangulation.
+        model: LinkModel
+            Per-link model in time used to correct the gradients before
+        unwrapping.
+        """
+        self._solver_space = solver_space
+        self._solver_time = solver_time
+        self._settings = settings if settings else Settings()
+        self._link_model = link_model
+
+        if link_model is not None:
+            errmsg = "Not implemented yet."
+            raise NotImplementedError(errmsg)
+
+    @property
+    def npoints(self) -> int:
+        """Number of points in the spatial network."""
+        return self._solver_space.npoints
+
+    @property
+    def nlinks(self) -> int:
+        """Number of links in the spatial network."""
+        return self._solver_space.nedges
+
+    @property
+    def nepochs(self) -> int:
+        """Number of points in the temporal network."""
+        return self._solver_time.npoints
+
+    @property
+    def nifgs(self) -> int:
+        """Number of links in the temporal network."""
+        return self._solver_time.nedges
+
+    @property
+    def settings(self) -> Settings:
+        """Retrieve settings for the workflow."""
+        return self._settings
+
+    @property
+    def link_model(self) -> LinkModelInterface | None:
+        """Retrieve the link model for the workflow."""
+        return self._link_model
+
+    def unwrap_cube(self, wrap_data: np.ndarray) -> np.ndarray:
+        """Unwrap a 3D cube of data.
+
+        Parameters
+        ----------
+        wrap_data: np.ndarray
+            2D array of size (nslc, npoints) or (nifg, npoints).
+
+        Returns
+        -------
+        uw_data: np.ndarray
+            2D float32 array of size (nifg, npoints).
+        """
+        if wrap_data.ndim != 2:
+            errmsg = f"Input data has more than two dimension - {wrap_data.ndim}"
```

Comment:
[NB] It could be that `wrap_data.ndim == 1`, in which case the error message will not make sense.

Reply:
Done